### PR TITLE
signal: add hooks for calling fatal error handlers from non-envoy signal handlers.

### LINKED ISSUE #11984: Allow ScopeTrackedObject crash logging without Envoy signal handler
We have funcionality for dumping active sessions on crash, as part of issue #7300, but it isn't usable except through Envoy's signal handler: the session dump is only accessible via FatalErrorHandlerInterface, and Envoy::Event::DispatcherImpl won't register its fatal error handler unless Envoy is compiled with ENVOY_HANDLE_SIGNALS. I'd like to add an interface to call into the fatal error handlers from signal handlers other than Envoy's.

### PR DESCRIPTION
**Pull request by mkbehr**
Commit Message: Add hooks for calling fatal error handlers from non-Envoy signal handlers.
Additional Description:
Move register/removeFatalErrorHandler from SignalAction into a new FatalErrorHandler namespace, and add a new function, callFatalErrorHandlers, which runs the registered error handlers. This makes the crash logging from issue #7300 available for builds that don't use ENVOY_HANDLE_SIGNALS, as long as they do use ENVOY_OBJECT_TRACE_ON_DUMP.
Risk Level: Low
Testing: bazel test //test/...
Docs Changes: n/a
Release Notes: yes
Fixes #11984

### CODE REVIEW

**Comment by mkbehr** [comment]
Trying to replicate the presubmit failures on my machine. I suspect the mock ends up allocating some memory, but having trouble confirming.

**Comment by mkbehr** [comment]
Replicated the failure under Docker, so this should fix it.

**Comment by dio** [comment]
Sorry, @mkbehr could you check CI?

**Comment by mkbehr** [comment]
Asan CI should be fixed. Clang-tidy is taking a few tries since I'm not able to run it locally and it seems to be picking up old errors - hopefully this run will work.

**Comment by mkbehr** [comment]
CI is passing now - can you have a look?

File: source/common/signal/fatal_error_handler.cc [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+    fatal_error_handlers.store(list, std::memory_order_release);
+  }
+#else
+  UNREFERENCED_PARAMETER(handler);
+#endif
+}
+
+void callFatalErrorHandlers(std::ostream& os) {
+  FailureFunctionList* list = fatal_error_handlers.exchange(nullptr, std::memory_order_relaxed);
+  if (list) {
```
Comment (dio) [inline source/common/signal/fatal_error_handler.cc:64]: Optional, prefer to check it explicitly: `if (list != nullptr)`

**Comment by dio** [comment]
Syncing to head will help with the conflicting files.

File: source/common/signal/fatal_error_handler.cc (hunk 1 above)
Comment (mkbehr) [reply source/common/signal/fatal_error_handler.cc]: Done.

**Comment by mkbehr** [comment]
CI failure doesn't look related to this PR: it just timed out trying to compile a test.

**Comment by mkbehr** [comment]
@zuercher, have you had a chance to look at this?

File: source/common/signal/fatal_error_handler.cc [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+namespace {
+
+ABSL_CONST_INIT static absl::Mutex failure_mutex(absl::kConstInit);
+// Since we can't grab the failure mutex on fatal error (snagging locks under
+// fatal crash causing potential deadlocks) access the handler list as an atomic
+// operation, to minimize the chance that one thread is operating on the list
+// while the crash handler is attempting to access it.
+// This basically makes edits to the list thread-safe - if one thread is
+// modifying the list rather than crashing in the crash handler due to accessing
+// the list in a non-thread-safe manner, it simply won't log crash traces.
```
[the remark this answers is not in the document]
Comment (zuercher) [reply source/common/signal/fatal_error_handler.cc:20]: Oh yay, github lost my edited comment. I meant to say:

I know you didn't write this comment, but can you clarify it. I think the important point is that the crash handler always sees with a valid handler list pointer, or while the list is being updated, nullptr. And likewise the modification code either sees a valid handler list pointer or a nullptr if we're mid-crash.

[the remark this answers is not in the document]
Comment (mkbehr) [reply source/common/signal/fatal_error_handler.cc]: Yeah, that sounds right to me. How's this version?

**Comment by mkbehr** [comment]
A few presubmit failures in the latest push, but they look unrelated.

**Comment by zuercher** [comment]
Looks good, but you'll need to merge master again.

**Comment by zuercher** [comment]
@lizan do you know why clang-tidy is complaining about the null pointer dereference despite the `NOLINTNEXTLINE` directive?

```
test/common/signal/signals_test.cc:57:22: error: Dereference of null pointer (loaded from variable 'nasty_ptr') [clang-analyzer-core.NullDereference,-warnings-as-errors]
        *(nasty_ptr) = 0;
```

**Comment by lizan** [comment]
@zuercher https://github.com/envoyproxy/envoy/pull/12372

**Comment by zuercher** [comment]
Thanks @lizan!

@mkbehr If you merge master once more it should resolve this last test error. Sorry for the churn.

**Comment by zuercher** [comment]
Ugh. Looks like you caught bad merge that landed on master.

**Comment by mkbehr** [comment]
Okay, trying again.

**Comment by mkbehr** [comment]
Looks like I caught yet another bad merge. Trying again.

As a heads up, Friday will be my last day working on Envoy, so hopefully we can get this merged by then.

**Comment by mkbehr** [comment]
Okay, looks like I found a good merge.

**Pull request merged**